velox/functions/prestosql/types/fuzzer_utils/TimestampWithTimeZoneInputGenerator.cpp

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/prestosql/types/fuzzer_utils/TimestampWithTimeZoneInputGenerator.h"
+
+#include "velox/common/fuzzer/Utils.h"
+#include "velox/functions/prestosql/types/TimestampWithTimeZoneType.h"
+#include "velox/type/Variant.h"
+#include "velox/type/tz/TimeZoneMap.h"
+
+namespace facebook::velox::fuzzer {
+TimestampWithTimeZoneInputGenerator::TimestampWithTimeZoneInputGenerator(
+    const size_t seed,
+    const double nullRatio)
+    : AbstractInputGenerator(
+          seed,
+          TIMESTAMP_WITH_TIME_ZONE(),
+          nullptr,
+          nullRatio),
+      timeZoneIds_(tz::getTimeZoneIDs()) {}
+
+variant TimestampWithTimeZoneInputGenerator::generate() {
+  if (coinToss(rng_, nullRatio_)) {
+    return variant::null(type_->kind());
+  }
+
+  int16_t timeZoneId = timeZoneIds_
+      [boost::random::uniform_int_distribution<size_t>()(rng_) %
+       timeZoneIds_.size()];
+
+  return pack(
+      boost::random::uniform_int_distribution<int64_t>()(rng_), timeZoneId);
```

Comment:
Would we want to occasionally generate invalid TimestampWithTimezone to test invalid cornercases? (Similar to JsonInputGenerator::makeRandomVariation().)

Reply:
I'm not sure if we need that given that TSwTZ is an internal type (it can only occur if it's the output of another UDF, it cannot be part of the input).

The only way a TSwTZ could be invalid is if its timeZoneId is some non-existent value. The only way an invalid TSwTZ could be produced is if a UDF is producing them. As long as we have a UDF that is capable of producing them (from_unixtimestamp comes to mind) included in fuzzing we will cover such cases in fuzzing without having it explicitly here.

If we have a UDF (like json_parse with JsonInputGenerator) where we explicitly want to test invalid values we can
extend this to support it.

Reply:
Actually, looking closer at JsonInputGenerator it produces either a Varchar or a Json type, JsonInputGenerator::makeRandomVariation() is used when outputting a Varchar that is the input to json_parse.  So a better analog would be some yet to be written TimeZoneIdInputGenerator for from_unixtimestamp where we would want to occasionally return an invalid time zone ID for the SmallInt argument.